### ISSUE TITLE
Object Not Set to an Instance of An Object : ApiClient.GetUserProfile() line 112 : {Development}

### ISSUE THREAD

**Issue by trihugger** [enkodellc/blazorboilerplate#392]
Hello,

I've encountered this error which is very odd and I can't seem to pin point the issue. Let me bullet point my steps:

- Create a DataModel in \Shared\BlazorBoilerplate.Infrastructure\Storage\DataModels "see code below"
- Add to ApplicationContext DB Set to the datamodel
`        public DbSet<TenantSetting> TenantSettings { get; set; }

        public DbSet<ApiLogItem> ApiLogs { get; set; }
        public DbSet<UserProfile> UserProfiles { get; set; }
        public DbSet<Todo> Todos { get; set; }
        **public DbSet<PayoffLoan> PayoffLoans { get; set; }**
        public DbSet<Message> Messages { get; set; }
        private IUserSession _userSession { get; set; }
        public DbSet<DbLog> Logs { get; set; }`
- Added Migration
`dotnet ef --startup-project ../BlazorBoilerplate.Server/ migrations add TitlesDBset1 -c ApplicationDbContext --verbose --no-build --configuration Debug`
- Build and Run application

Get the error that "Object is not set to an instance of an object". When I go to the console to see the error it points to the GetUserprofile. Full error message:
`Error: System.Exception: Unable to locate metadata resource for: http://localhost:53415/api/data/
 ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at Breeze.Sharp.MetadataStore.UpdateStructuralTypeFromJNode(JNode jNode, Boolean isFromServer)
   at Breeze.Sharp.MetadataStore.<>c__DisplayClass53_0.<DeserializeFrom>b__2(JNode jn)
   at Breeze.Sharp.Core.EnumerableFns.ForEach[T](IEnumerable`1 items, Action`1 action)
   at Breeze.Sharp.MetadataStore.DeserializeFrom(JNode jNode, Boolean isFromServer)
   at Breeze.Sharp.MetadataStore.ImportMetadata(JNode jNode, Boolean isFromServer)
   at Breeze.Sharp.MetadataStore.ImportMetadata(String metadata, Boolean isFromServer)
   at Breeze.Sharp.MetadataStore.FetchMetadata(DataService dataService, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at Breeze.Sharp.MetadataStore.FetchMetadata(DataService dataService, CancellationToken cancellationToken)
   at Breeze.Sharp.EntityManager.FetchMetadata(CancellationToken cancellationToken, DataService dataService)
   at Breeze.Sharp.EntityManager.FetchMetadata(DataService dataService)
   at Breeze.Sharp.EntityManager.ExecuteQuery(EntityQuery query, CancellationToken cancellationToken)
   at Breeze.Sharp.EntityManager.ExecuteQuery[T](EntityQuery`1 query, CancellationToken cancellationToken)
   at BlazorBoilerplate.Shared.Services.ApiClient.GetUserProfile() in C:\...\blazorboilerplate-ReDev\src\Shared\BlazorBoilerplate.Shared\Services\ApiClient.cs:line 112
   at BlazorBoilerplate.Shared.Services.AppState.GetUserProfile() in C:\...\blazorboilerplate-ReDev\src\Shared\BlazorBoilerplate.Shared\Services\AppState.cs:line 68
   at BlazorBoilerplate.Theme.Material.Shared.Layouts.MainLayout.OnInitializedAsync() in C:\...\blazorboilerplate-ReDev\src\Shared\Modules\BlazorBoilerplate.Theme.Material\Shared\Layouts\MainLayout.razor:line 77
   at Microsoft.AspNetCore.Components.ComponentBase.RunInitAndSetParametersAsync()
   at Microsoft.AspNetCore.Components.RenderTree.Renderer.GetErrorHandledTask(Task taskToHandle)`

When I put a breakpoint i see it returns ApplicationUser as a null object. 

Just to be clear I did no changes to ApplicationUser or GetUserProfile, in fact I have started with a clean development copy and added just these changes to compare.

I also, checked with and without migration to see if it was a migration error, as well as xref the field names to see if it was creating any conflict with either Datamodels (ApplicationUser/UserProfile).

Any help appreciate.

Thank you,

Jeff

DataModel:
`namespace BlazorBoilerplate.Infrastructure.Storage.DataModels
{
    public partial class PayoffLoan : IAuditable, ISoftDelete
    {
        [Key]
        public long Id { get; set; }

        [Required]
        [MaxLength(15)]
        public string AccountNumber { get; set; }

        [Required]
        public DateTime PaidDate { get; set; }
        [MaxLength(100)]
        public string Product { get; set; }
        [Required]
        [MaxLength(100)]
        public string Borrower { get; set; }
        [MaxLength(100)]
        public string CoBorrower { get; set; }
        [Required]
        [MaxLength(17)]
        public string VIN { get; set; }

        [Required]
        public double PayOffAmount { get; set; }

        public bool ChargeOff { get; set; }
        public bool Recovered { get; set; }
        public bool Perfected { get; set; }
        public bool Paper { get; set; }

        [Required]
        public int ReleaseType { get; set; }
        public DateTime ReleaseDate { get; set; }

        [MaxLength(100)]
        public string Processor { get; set; }
        [MaxLength(100)]
        public string Approver { get; set; }
        [MaxLength(100)]
        public string Reviewer { get; set; }

        public int ReleaseStatus { get; set; }
    }
}`

**Comment by GioviQ** [comment]
Did you read this https://blazor-boilerplate.readthedocs.io/en/latest/quickstarts/breeze_sharp_with_blazor.html?

For every entity you have defined, you have to define a DTO entity.

**Comment by trihugger** [comment]
Hi GioviQ,

I did, that was very helpful. Originally I had a DTO as well, so since I started from scratch I deleted but here's the code.

Again same error.

Here's the DTO code:

`using System;
using System.Collections.Generic;
using BlazorBoilerplate.Shared.Dto.Db;
using Breeze.Sharp;

namespace BlazorBoilerplate.Shared.Dto.Db
{
    public partial class PayoffLoanDto : BaseEntity
    {
        #region DataModel Fields
        public Int64 Id
        {
            get { return GetValue<Int64>(); }
            set { SetValue(value); }
        }

        public String AccountNumber
        {
            get { return GetValue<String>(); }
            set { SetValue(value); }
        }

        public DateTime PaidDate
        {
            get { return GetValue<DateTime>(); }
            set { SetValue(value); }
        }
        public String Product
        {
            get { return GetValue<String>(); }
            set { SetValue(value); }
        }
        public String Borrower
        {
            get { return GetValue<String>(); }
            set { SetValue(value); }
        }
        public String CoBorrower
        {
            get { return GetValue<String>(); }
            set { SetValue(value); }
        }
        public String VIN
        {
            get { return GetValue<String>(); }
            set { SetValue(value); }
        }

        public Double PayOffAmount
        {
            get { return GetValue<Double>(); }
            set { SetValue(value); }
        }

        public Boolean ChargeOff
        {
            get { return GetValue<Boolean>(); }
            set { SetValue(value); }
        }
        public Boolean Recovered
        {
            get { return GetValue<Boolean>(); }
            set { SetValue(value); }
        }
        public Boolean Perfected
        {
            get { return GetValue<Boolean>(); }
            set { SetValue(value); }
        }
        public Boolean Paper
        {
            get { return GetValue<Boolean>(); }
            set { SetValue(value); }
        }

        public ReleaseType ReleaseType
        {
            get { return GetValue<ReleaseType>(); }
            set { SetValue(value); }
        }
        public DateTime ReleaseDate
        {
            get { return GetValue<DateTime>(); }
            set { SetValue(value); }
        }

        public String Processor
        {
            get { return GetValue<String>(); }
            set { SetValue(value); }
        }
        public String Approver
        {
            get { return GetValue<String>(); }
            set { SetValue(value); }
        }
        public String Reviewer
        {
            get { return GetValue<String>(); }
            set { SetValue(value); }
        }

        public LoanReleaseStatus ReleaseStatus
        {
            get { return GetValue<LoanReleaseStatus>(); }
            set { SetValue(value); }
        }

        #endregion

        #region BPP Fields
        //Add the following for the purpose of IAuditable and ISoftDelete functionality
        public Guid? CreatedById
        {
            get { return GetValue<Guid?>(); }
            set { SetValue(value); }
        }

        public DateTime CreatedOn
        {
            get { return GetValue<DateTime>(); }
            set { SetValue(value); }
        }

        public Boolean IsDeleted
        {
            get { return GetValue<Boolean>(); }
            set { SetValue(value); }
        }

        public Guid? ModifiedById
        {
            get { return GetValue<Guid?>(); }
            set { SetValue(value); }
        }

        public DateTime ModifiedOn
        {
            get { return GetValue<DateTime>(); }
            set { SetValue(value); }
        }
        public ApplicationUser CreatedBy
        {
            get { return GetValue<ApplicationUser>(); }
            set { SetValue(value); }
        }

        public ApplicationUser ModifiedBy
        {
            get { return GetValue<ApplicationUser>(); }
            set { SetValue(value); }
        }
        #endregion
    }

    public enum ReleaseType
    {
        Immediate,
        [Display(Name = "10 Days")]
        TenDays
    }

    public enum LoanReleaseStatus
    {
        Pending = 1, //All loans that come in - Step1
        Recovered = 2, //Loans waiting on MAG for Release - Step 2 from 1
        Approved = 3, //Approved by Loan Ops to Process - Step 2
        Manual = 4, //Loan that must be processed Manually by a person. E.g. Paper Titles - Step 2
        Processed = 5, //Processed by Robot or Human - Step 3
        //Chaged Off loans go straight to Complete Status
        Completed = 6 //Final Step - Step 4
    }
}`

**Comment by trihugger** [comment]
Actually found the issue, I changed the name of the DTO class to include the name Dto e.g. PayoffLoan and PayoffLoanDto. On a hunch I removed that and it worked perfectly.

I Like Breeze, I just need to learn all the stuff around it.

On a side note, if anyone is looking I had changed (trying to reverse engineer my issue) reverted the enum ReleaseType and LoanReleaseStatus to Int32 in the Datamodel, but it has to be of the same type.

Giovianni, just wanted to thank you for all the work with the development restructuring and implementing Breeze, like you pointed out it has its drawbacks, but it seems to streamline the process.

One question on DataAnnotation, is it best practice to place your DataAnnotations in both DTO and Data Model, or in just one side or the other? E.g. DataFormat, Required, etc.
